### PR TITLE
Refactor/Replace websockets library with aiohttp library - Kraken

### PR DESCRIPTION
**Pull request by mhrvth**
**Before submitting this PR, please make sure**:

- [x] Your code builds clean without any errors or warnings
- [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:

websockets replaced with web_assistant (aiohttp)

**Tests performed by the developer**:

Ran the unit tests

### CODE REVIEW

**Comment by rapcmia** [comment]
Partially tested and getting `unhandled error in background task: fetch_trading_pairs() missing...` upon opening the client. Still tried to run Pure market-making getting `unexpected error in user stream..` and `unexpected errors..` Tested on docker and source build on both Windows and Linux/AWS with the same results.

image

Error after you open the client
```
2021-12-14 22:56:05,070 - 5045 - hummingbot.core.utils.async_utils - ERROR - Unhandled error in background task: fetch_trading_pairs() missing 1 required positional argument: 'client'
Traceback (most recent call last):
  File "/home/rapcmia/github/4907/hummingbot/core/utils/async_utils.py", line 9, in safe_wrapper
    return await c
  File "/home/rapcmia/github/4907/hummingbot/core/utils/trading_pair_fetcher.py", line 58, in fetch_all
    safe_ensure_future(self.call_fetch_pairs(module.fetch_trading_pairs(**args), conn_setting.name))
TypeError: fetch_trading_pairs() missing 1 required positional argument: 'client'
```

Error after `start` command
```
2021-12-14 23:02:58,674 - 6986 - hummingbot.connector.exchange.kraken.kraken_api_order_book_data_source - ERROR - Unexpected error. 
Traceback (most recent call last):
  File "/home/rapcmia/github/4907/hummingbot/connector/exchange/kraken/kraken_api_order_book_data_source.py", line 277, in listen_for_order_book_snapshots
    client = await self._get_rest_assistant()
  File "/home/rapcmia/github/4907/hummingbot/connector/exchange/kraken/kraken_api_order_book_data_source.py", line 66, in _get_rest_assistant
    self._rest_assistant = await self._api_factory.get_rest_assistant()
AttributeError: 'NoneType' object has no attribute 'get_rest_assistant'
```

```
hummingbot.connector.exchange.kraken.kraken_api_order_book_data_source - ERROR - Unexpected error with WebSocket connection. Retrying after 30 seconds...
Traceback (most recent call last):
  File "/home/rapcmia/github/4907/hummingbot/connector/exchange/kraken/kraken_api_order_book_data_source.py", line 246, in listen_for_order_book_diffs
    msg_dict = {"trading_pair": convert_from_exchange_trading_pair(msg[-1]),
KeyError: -1
2021-12-14 23:02:59,804 - 6986 - hummingbot.connector.exchange.kraken.kraken_exchange - ERROR - Unexpected error in user stream listener loop.
```

Crosscheck testbot running on the dev branch, Kraken with the same config seems to be running ok. Also, recompile/install (3/3) getting the same errors

Steps to reproduce:
1. Open the client 
2. Create Pure market making strategy and start the bot

Attached are log and config:
[pmm-test.zip](https://github.com/CoinAlpha/hummingbot/files/7712786/pmm-test.zip)

**Comment by rapcmia** [comment]
The reported issue has been fixed on the latest commit. Upon testing order-filled events, the client seems to be having issues with` history`. When there is a filled order, the bot doesn't seem to record it. On the screenshot below, you would notice the filled order has been added on the logs, and the connector's trade history shows the data but not on the client's `history`, it doesn't have a record on CSV/SQLite file

Screenshot 2021-12-15 223159
image

You would also notice that there are executed trades that are captured on the connector's trade history but not on the client. This has been tested on both` local and AWS instances` getting the same results. On the screenshot below, the order_id `1639578587001079 ` has been created and cancelled yet the connector's trade history shows it was filled.

Screenshot 2021-12-15 223842
```
# Creating orders 
2021-12-15 22:29:47,000 - 3880 - hummingbot.strategy.pure_market_making.pure_market_making - INFO - (BTC-USDC) Creating 1 bid orders at (Size, Price): ['0.00029936 BTC, 47716.49 USDC']
2021-12-15 22:29:47,001 - 3880 - hummingbot.strategy.pure_market_making.pure_market_making - INFO - (BTC-USDC) Creating 1 ask orders at (Size, Price): ['0.0003 BTC, 47745.13 USDC']

# Orders created 
2021-12-15 22:29:47,213 - 3880 - hummingbot.connector.exchange.kraken.kraken_exchange - INFO - Created OrderType.LIMIT_MAKER sell order sell-BTC-USDC-1639578587001206 for 0.00030000 BTC-USDC.
2021-12-15 22:29:47,229 - 3880 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1639578587.0, "type": "OrderType.LIMIT_MAKER", "trading_pair": "BTC-USDC", "amount": "0.00030000", "price": "47745.13", "order_id": "sell-BTC-USDC-1639578587001206", "exchange_order_id": null, "leverage": 1, "position": "NILL", "event_name": "SellOrderCreatedEvent", "event_source": "kraken"}
2021-12-15 22:29:47,430 - 3880 - hummingbot.connector.exchange.kraken.kraken_exchange - INFO - Created OrderType.LIMIT_MAKER buy order buy-BTC-USDC-1639578587001079 for 0.00029936 BTC-USDC.
2021-12-15 22:29:47,446 - 3880 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1639578587.0, "type": "OrderType.LIMIT_MAKER", "trading_pair": "BTC-USDC", "amount": "0.00029936", "price": "47716.49", "order_id": "buy-BTC-USDC-1639578587001079", "exchange_order_id": null, "leverage": 1, "position": "NILL", "event_name": "BuyOrderCreatedEvent", "event_source": "kraken"}

# Orders cancelled 
2021-12-15 22:30:47,000 - 3880 - hummingbot.strategy.pure_market_making.pure_market_making - INFO - (BTC-USDC) Cancelling the limit order buy-BTC-USDC-1639578587001079. [clock=2021-12-15 14:30:47+00:00]
2021-12-15 22:30:47,001 - 3880 - hummingbot.strategy.pure_market_making.pure_market_making - INFO - (BTC-USDC) Cancelling the limit order sell-BTC-USDC-1639578587001206. [clock=2021-12-15 14:30:47+00:00]
2021-12-15 22:30:47,295 - 3880 - hummingbot.connector.exchange.kraken.kraken_exchange - INFO - Successfully cancelled order sell-BTC-USDC-1639578587001206.
2021-12-15 22:30:47,310 - 3880 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1639578647.0, "order_id": "sell-BTC-USDC-1639578587001206", "exchange_order_id": null, "event_name": "OrderCancelledEvent", "event_source": "kraken"}
2021-12-15 22:30:47,491 - 3880 - hummingbot.connector.exchange.kraken.kraken_exchange - INFO - Successfully cancelled order buy-BTC-USDC-1639578587001079.
2021-12-15 22:30:47,506 - 3880 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1639578647.0, "order_id": "buy-BTC-USDC-1639578587001079", "exchange_order_id": null, "event_name": "OrderCancelledEvent", "event_source": "kraken"}
```

I was also able to capture an instance of partially filled orders where the bot was not able to add to the logs each filled event.
Screenshot 2021-12-15 232727

Steps to reproduce:
1. Setup Pure market making for Kraken
2. Set tight spreads and monitor every filled event and connector's trade history
3. Run `history` command

Attached logs and config files:
[pmm-test.zip](https://github.com/CoinAlpha/hummingbot/files/7720744/pmm-test.zip)

**Comment by RC-13** [comment]
Got the same experience with filled trade
image
image
Upon checking the logs, there was no completion event due to the following error:
```python
2021-12-16 03:44:30,250 - 11965 - hummingbot.connector.exchange.kraken.kraken_exchange - INFO - The market sell order sell-DOGE-USDT-1639626249002327 has completed according to order status API.
2021-12-16 03:44:30,260 - 11965 - hummingbot.strategy.pure_market_making.pure_market_making - INFO - (DOGE-USDT) Maker sell order sell-DOGE-USDT-1639626249002327 (60.00000000 DOGE @ 0.18092 USDT) has been completely filled. [clock=2021-12-16 03:44:30+00:00]
2021-12-16 03:44:30,260 - 11965 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1639626270.0, "order_id": "sell-DOGE-USDT-1639626249002327", "base_asset": "DOGE", "quote_asset": "USDT", "fee_asset": "USDT", "base_asset_amount": "60.00000000", "quote_asset_amount": "10.8552000000000", "fee_amount": "0.01736", "order_type": "OrderType.LIMIT_MAKER", "exchange_order_id": null, "event_name": "SellOrderCompletedEvent", "event_source": "kraken"}
2021-12-16 03:44:34,322 - 11965 - hummingbot.connector.exchange.kraken.kraken_api_user_stream_data_source - ERROR - Unexpected error with Kraken WebSocket connection. Retrying after 30 seconds...
Traceback (most recent call last):
  File "/home/ubuntu/kraken/hummingbot/connector/exchange/kraken/kraken_api_user_stream_data_source.py", line 97, in listen_for_user_stream
    self._current_auth_token = await self.get_auth_token()
  File "/home/ubuntu/kraken/hummingbot/connector/exchange/kraken/kraken_api_user_stream_data_source.py", line 88, in get_auth_token
    return response_json["result"]["token"]
KeyError: 'result'
```
[logs_conf_pure_mm_1.log](https://github.com/CoinAlpha/hummingbot/files/7724429/logs_conf_pure_mm_1.log)

**Comment by rxlxrxsx** [comment]
- Intermittently getting this error while the bot is running. This error is only visible in the log file.

```python
2021-12-16 10:29:56,893 - 7446 - hummingbot.connector.exchange.kraken.kraken_order_book_tracker - NETWORK - Unexpected error while fetching last trade price.
Traceback (most recent call last):
  File "/Users/reiniersalazar/hummingbot4/hummingbot/core/data_type/order_book_tracker.py", line 161, in _update_last_trade_prices_loop
    last_prices = await self._data_source.get_last_traded_prices(**args)
TypeError: get_last_traded_prices() got an unexpected keyword argument 'domain'
```

Logs and config: [kraken_aiohttp.zip](https://github.com/CoinAlpha/hummingbot/files/7724449/kraken_aiohttp.zip)

**Comment by mhrvth** [comment]
Looking at the old code the domain exception had to be happening for a while unnoticed

**Comment by mhrvth** [comment]
For the missing result key problem I'm getting the correct response and no exception

image

**Review by petioptrv — CHANGES_REQUESTED** [review]
I think the template for this refactor was Gate.io. However, Gate.io is particular in that it implements an entire separate websocket class. That class existed in Gate.io connector since before the `aiohttp` refactor and was outside of the scope of the refactor task to remove it, so it still remains there and instead, the only thing that was refactored was that class' use of `aiohttp` and `websockets`.

That being said, adapting that code has led to some major changes to the logic and I am certain that the bugs that QA is observing are hiding somewhere in there. Also, there are now some sections of the code that are not relevant to Kraken (but were relevant to Gate.io), like the `"spot.ping"` websocket endpoint and the logic around it.

That being said, I think the easiest and cleanest way to resolve this is to change the adaptation to model the dydx connector instead, which is much closer to the "default" implementation, and which will require much less changes to the already existing code in Kraken.

Basically, the goal is to simply replace in the code where either `aiohttp` or `websockets` is used with the `WebAssistantsFactory` objects

File: hummingbot/connector/exchange/kraken/kraken_api_user_stream_data_source.py [hunk 1]
```diff
@@ -35,12 +36,13 @@ def logger(cls) -> HummingbotLogger:
             cls._krausds_logger = logging.getLogger(__name__)
         return cls._krausds_logger
 
-    def __init__(self, throttler: AsyncThrottler, kraken_auth: KrakenAuth):
+    def __init__(self, throttler: AsyncThrottler, kraken_auth: KrakenAuth, api_factory: Optional[WebAssistantsFactory] = None):
```
Comment (petioptrv) [inline hummingbot/connector/exchange/kraken/kraken_api_user_stream_data_source.py:39]: this line is over 120 characters

Comment (petioptrv) [reply hummingbot/connector/exchange/kraken/kraken_api_user_stream_data_source.py:39]:
```suggestion
    def __init__(
        self, throttler: AsyncThrottler, kraken_auth: KrakenAuth, api_factory: Optional[WebAssistantsFactory] = None
    ):
```

File: hummingbot/connector/exchange/kraken/kraken_api_order_book_data_source.py [hunk 2]
```diff
@@ -45,42 +47,55 @@ def logger(cls) -> HummingbotLogger:
             cls._kraobds_logger = logging.getLogger(__name__)
         return cls._kraobds_logger
 
-    def __init__(self, throttler: Optional[AsyncThrottler] = None, trading_pairs: List[str] = None):
+    def __init__(self, throttler: Optional[AsyncThrottler] = None, trading_pairs: List[str] = None, api_factory: Optional[WebAssistantsFactory] = None):
```
[the remark this answers is not in the document]
Comment (petioptrv) [reply hummingbot/connector/exchange/kraken/kraken_api_order_book_data_source.py:48]:
```suggestion
    def __init__(
        self,
        throttler: Optional[AsyncThrottler] = None,
        trading_pairs: List[str] = None,
        api_factory: Optional[WebAssistantsFactory] = None,
    ):
```

File: hummingbot/connector/exchange/kraken/kraken_api_order_book_data_source.py [hunk 3]
```diff
@@ -96,81 +110,72 @@ async def get_snapshot(
...
+            if len(response_json["error"]) > 0:
+                raise IOError(f"Error fetching Kraken market snapshot for {original_trading_pair}. "
+                              f"Error is {response_json['error']}.")
+            data: Dict[str, Any] = next(iter(response_json["result"].values()))
+            data = {"trading_pair": trading_pair, **data}
+            data["latest_update"] = max([*map(lambda x: x[2], data["bids"] + data["asks"])], default=0.)
+
+            # Need to add the symbol into the snapshot message for the Kafka message queue.
+            # Because otherwise, there'd be no way for the receiver to know which market the
+            # snapshot belongs to.
```
Comment (petioptrv) [inline hummingbot/connector/exchange/kraken/kraken_api_order_book_data_source.py:135]:
```suggestion
```

File: hummingbot/connector/exchange/kraken/kraken_order_book_tracker.py [hunk 4]
```diff
@@ -45,7 +48,7 @@ def __init__(self,
     @property
     def data_source(self) -> OrderBookTrackerDataSource:
         if not self._data_source:
-            self._data_source = KrakenAPIOrderBookDataSource(trading_pairs=self._trading_pairs)
+            self._data_source = KrakenAPIOrderBookDataSource(trading_pairs=self._trading_pairs, api_factory=self._api_factory)
```
Comment (petioptrv) [inline hummingbot/connector/exchange/kraken/kraken_order_book_tracker.py:51]: this line is over 120 characters
```suggestion
            self._data_source = KrakenAPIOrderBookDataSource(
                trading_pairs=self._trading_pairs, api_factory=self._api_factory
            )
```

**Review by petioptrv — CHANGES_REQUESTED** [review]
I've added suggestions for line-breaks to avoid line lengths over 120 characters. Other than that, looks great and will be ready to approve once those are addressed.

@CoinAlpha/qa-team, these changes will have no impact on the tests you are currently performing.

**Review by RC-13 — APPROVED** [review]
Confirmed that the reported issue has been addressed in the latest commit. LGTM!

**Review by petioptrv — APPROVED** [review]
Changes LGTM!
